client/extensions/woocommerce/woocommerce-services/views/packages/package-dialog.js

```diff
@@ -0,0 +1,165 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { localize } from 'i18n-calypso';
+import { concat, difference, flatten, map } from 'lodash';
+import Gridicon from 'gridicons';
+
+/**
+ * Internal dependencies
+ */
+import EditPackage from './edit-package';
+import checkInputs from './modal-errors';
+import Dialog from 'components/dialog';
+import FormSectionHeading from 'components/forms/form-section-heading';
+import FormButton from 'components/forms/form-button';
+import inputFilters from './input-filters';
+import PredefinedPackages from './predefined-packages';
+import SegmentedControl from 'components/segmented-control';
+import { getPredefinedPackagesChangesSummary } from '../../state/packages/selectors';
+
+const AddPackageDialog = ( props ) => {
+	const {
+		siteId,
+		form,
+		predefinedPackagesSummary,
+		setModalErrors,
+		savePackage,
+		savePredefinedPackages,
+		dismissModal,
+		removePackage,
+		setAddMode,
+		updatePackagesField,
+		translate,
+	} = props;
+
+	const {
+		showModal,
+		mode,
+		packages,
+		packageSchema,
+		predefinedSchema,
+		packageData,
+	} = form;
+
+	const {
+		index,
+	} = packageData;
+
+	const customPackages = packages.custom;
+
+	const onSave = () => {
+		if ( 'add-predefined' === mode ) {
+			savePredefinedPackages( siteId );
+			return;
+		}
+
+		const editName = 'number' === typeof packageData.index ? customPackages[ packageData.index ].name : null;
+
+		//get reserved box names:
+		const boxNames = concat(
+			difference( customPackages.map( ( boxPackage ) => boxPackage.name ), [ editName ] ), //existing custom boxes
+			flatten( map( predefinedSchema, predef => map( predef, group => group.definitions ) ) ), //predefined boxes
+			[ 'individual' ] //reserved for items shipping in original packaging
+		);
+
+		const filteredPackageData = Object.assign( {}, packageData, {
+			name: inputFilters.string( packageData.name ),
+			inner_dimensions: inputFilters.dimensions( packageData.inner_dimensions ),
+			outer_dimensions: inputFilters.dimensions( packageData.outer_dimensions ),
+			box_weight: inputFilters.number( packageData.box_weight ),
+			max_weight: inputFilters.number( packageData.max_weight ),
+		} );
+
+		const errors = checkInputs( filteredPackageData, boxNames, packageSchema );
+		if ( errors.any ) {
+			updatePackagesField( siteId, filteredPackageData );
+			setModalErrors( siteId, errors );
+			return;
+		}
+
+		savePackage( siteId, filteredPackageData );
+	};
+
+	const onClose = () => ( dismissModal( siteId ) );
+	const onRemove = () => removePackage( siteId, index );
+
+	const switchMode = ( option ) => {
+		setAddMode( siteId, option.value );
+	};
+
+	const heading = 'edit' === mode ? translate( 'Edit package' ) : translate( 'Add a package' );
```

Comment:
You use add-predefined === mode and edit === mode and add-custom === mode quite a few times here. Please consider defining intermediate variables to increase readability and avoid breakage down the road if someone misses one.